src/tasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,381 @@
+        // for "0x1a, ". This is just a faster alternative to calling .ToString("x2") on every
+        // byte of the input file and then pushing that string through UTF8Encoding.
+        var lookup = new byte[256 * 6];
+        for (int i = 0; i < 256; i++)
+        {
+            string byteAsHex = i.ToString("x2");
+            char highOrderChar = BitConverter.IsLittleEndian ? byteAsHex[0] : byteAsHex[1];
+            char lowOrderChar = BitConverter.IsLittleEndian ? byteAsHex[1] : byteAsHex[0];
+            lookup[i * 6 + 0] = (byte)'0';
+            lookup[i * 6 + 1] = (byte)'x';
+            lookup[i * 6 + 2] = (byte)highOrderChar;
+            lookup[i * 6 + 3] = (byte)lowOrderChar;
+            lookup[i * 6 + 4] = (byte)',';
+            lookup[i * 6 + 5] = (byte)' ';
+        }
+
+        return lookup;
+    }
+
+    public abstract bool Emit(string destinationFile, Action<Stream> inputProvider);
+
+    private static Dictionary<string, int> symbolDataLen = new();
+
+    private static void GenerateBundledResourcePreallocationAndRegistration(string bundleRegistrationFunctionName, ICollection<(string registeredFilename, string resourceName, string culture, string? resourceSymbolName)> files, StreamWriter outputUtf8Writer)
+    {
+        StringBuilder preallocatedSource = new();
+
+        string assemblyTemplate = Utils.GetEmbeddedResource("mono-bundled-assembly.template");
+        string satelliteAssemblyTemplate = Utils.GetEmbeddedResource("mono-bundled-satellite-assembly.template");
+        string symbolDataTemplate = Utils.GetEmbeddedResource("mono-bundled-data.template");
+
+        var preallocatedResources = new StringBuilder();
+        var preallocatedAssemblies = new StringBuilder("MonoBundledResource *bundledAssemblyResources[] = { ");
+        var preallocatedSatelliteAssemblies = new StringBuilder("MonoBundledResource *bundledSatelliteAssemblyResources[] = { ");
+        var preallocatedData = new StringBuilder("MonoBundledResource *bundledDataResources[] = { ");
+        int assembliesCount = 0;
+        int satelliteAssembliesCount = 0;
+        int dataCount = 0;
+        foreach (var tuple in files)
+        {
+            string resourceId = tuple.registeredFilename;
+
+            // extern symbols
+            StringBuilder preallocatedResourceData = new();
+            preallocatedResourceData.AppendLine($"extern const unsigned char {tuple.resourceName}_data[];");
+            if (!string.IsNullOrEmpty(tuple.resourceSymbolName))
+            {
+                preallocatedResourceData.AppendLine($"extern const unsigned char {tuple.resourceSymbolName}_data[];");
+            }
+            preallocatedSource.AppendLine(preallocatedResourceData.ToString());
+
+            // Generate Preloaded MonoBundled*Resource structs
+            string preloadedStruct;
+            switch (GetFileType(tuple.registeredFilename)) {
```

Comment:
Maybe simplify this into looking at file extension and culture, no need to switch on a string. So drop use of switch and GetFileType and do:

```
if (!string.IsNullOrEmpty(culture))
{
    //Satellite assembly.
}
else if (destinationFileName.EndsWith(".dll"))
{
    //Assembly
}
else
{
    //Data
}
```

In both the assembly and satellite case you might do additional check that it is really a managed assembly, you can do that by applying similar logic as we have in AOT compiler to filter out unmanaged assemblies :

```
string assemblyPath = asmItem.GetMetadata("FullPath");
using var assemblyFile = File.OpenRead(assemblyPath);
using PEReader reader = new(assemblyFile, PEStreamOptions.Default);
if (!reader.HasMetadata)
{
    // Not a managed assembly.
}
```

This check is probably not as critical as it is in MonoAOTCompiler, since user would probably only pass in managed .dll's into the emit bundle task to begin with.

Reply:
Added the direct resource type check.

I believe at this step for generating source code based on Assembly/Satellite Assembly, the information passed are strings. Is there a way to check that an assembly is a managed assembly on the msbuild side?